Async monitored items

wbitesi opened the pull request:
Implemented client async functions for subscriptions/monitored items.
I tried to reduce the number of modifications of the original code by splitting each synchronous function in separate parts.

Longer explanation what and why it is implemented that way:

Split synchronous function (SFN) into the following parts:
a) prepare the call
b) execute the call synchronously
c) handle the call result

The asynchronous function (AFN) implements the same schema as SFN; it replace
the synchronous call by an asynchronous one.  It appends the c) part as
ClientAsyncCallback to the asynchronous call.  It also receives a user-defined
callback, that has to be executed after part c) of the AFN has run.

The difficulty lays in the memory management of parts a) and c).  In the
synchronous application, the function parameters and the local variables are
(except multithreading) safe and cannot be modified or deleted by the outside.

In the asynchronous case, part c) is executed after the AFN has returned.
Therefore, we must preserve the necessary state (parameters, local variables) on
the heap and manage the memory internally.  Since part c) is not user provided,
we have to ensure that everything is cleaned up appropriately.

I extended the CustomCallback to hold client-internal data along with a deleter
for that data.  The custom callback holds the usercall back and the ptr to its
userdata as well (which is called after the client-internal handler has run).

I updated the CustomCallback usage (read node async functions) according to the
extension.

basyskom-meerkoetter commented:
Please consider adding tests for your new code. See tests/client/ for what is already there.

wbitesi commented:
I'll take a look at the unit tests and try to add similar async based ones.

I'm not sure about the cppcheck warnings. When i run the cppcheck command (in travis_linux.sh) on my system using gcc-6.3.0 i dont get the warning as in the travis build. I dont understand why it isnt portable to have pointer arithmetic on void**. I thought that the c compiler knows the size of a void pointer.

Infinity95 commented:
Pointer arithmetic for void pointers is not legal in c or c++.
See https://stackoverflow.com/questions/3523145/pointer-arithmetic-for-void-pointer-in-c for an explanation.
Basically some compilers allow it as an extension and assume a size of one byte for void, which is why this is not portable.

wbitesi commented:
I have added test cases that check if subscriptions and monitored items are added/removed correctly.

The pull request should not fail (everything runs correctly), but the clang-format fails with "clang-format did not modify any files".

Pro commented:
If you do a new push and/or rebase onto master, the clang-format will be changed to allowed fail. It's currently under testing and should not block new PRs.
I will have a look at the false negative clang-format in the meantime.

wbitesi commented:
I wanted to ask if there is anything open for me to do with the pull request or if it is ready for merging and will be done if someone has time to do it?

jpfr commented on src/client/ua_client_internal.h:
```diff
@@ -111,16 +111,19 @@ void UA_Client_AsyncService_cancel(UA_Client *client, AsyncServiceCall *ac,
 
 void UA_Client_AsyncService_removeAll(UA_Client *client, UA_StatusCode statusCode);
 
+typedef void (*CustomCallbackDataDeleter)(void *);
```
Why do we need this guy?
If there is a timeout, we can just set a statuscode in the responseheader and call the "normal" callback.

wbitesi replied:
My motivation was to have the delete function in the custom callback to be able to free the memory if the handler isnt called. It looks like that the handler is always called (remove UA_Client_AsyncService_cancel) so i can remove it from the CustomCallback.

wbitesi replied:
I have replaced the deleter with a flag to distinguish between "async customcallbacks) on the heap and (sync customcallbacks) on the stack.

jpfr commented on src/client/ua_client_internal.h:
```diff
@@ -168,6 +171,31 @@ struct UA_Client {
     UA_Boolean pendingConnectivityCheck;
 };
 
+static UA_INLINE CustomCallback *CustomCallback_new(void) {
```
You can use UA_calloc to get zeroed memory.
So we probably don't need this method.

wbitesi replied:
I'll remove this method and replace it with UA_callocs

wbitesi replied:
done

jpfr commented on src/client/ua_client_subscriptions.c:
```diff
@@ -29,39 +70,62 @@ UA_Client_Subscriptions_create(UA_Client *client,
+    if (!cc) {
         response.responseHeader.serviceResult = UA_STATUSCODE_BADOUTOFMEMORY;
-        return response;
+        goto cleanup;
```
You can immediately return here.
There is no callback to clean up.

jpfr commented on src/client/ua_client_subscriptions.c:
```diff
@@ -29,39 +70,62 @@ UA_Client_Subscriptions_create(UA_Client *client,
+    __Subscriptions_create_handler(client, cc, 0, &response);
 
+    return response;
+cleanup:
```
The goto is then used in only one place.
Move up.

jpfr commented on src/client/ua_client_subscriptions.c:
```diff
@@ -29,39 +70,62 @@ UA_Client_Subscriptions_create(UA_Client *client,
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    CustomCallback *cc = CustomCallback_new();
+    if (!cc) {
+        retval = UA_STATUSCODE_BADOUTOFMEMORY;
```
There is no callback to remove.
Directly return.

jpfr commented on src/client/ua_client_subscriptions.c:
```diff
@@ -29,39 +70,62 @@ UA_Client_Subscriptions_create(UA_Client *client,
+                                    __Subscriptions_create_handler, &UA_TYPES[UA_TYPES_CREATESUBSCRIPTIONRESPONSE], cc,
+                                    requestId);
+
+cleanup:
```
The goto is used in only one place.
Move up.

jpfr commented on src/client/ua_client_subscriptions.c:
```diff
@@ -111,61 +210,134 @@ UA_Client_Subscription_deleteInternal(UA_Client *client, UA_Client_Subscription
             /* Something was wrong, reinsert the subscription in the list */
             if (subs[i])
                 LIST_INSERT_HEAD(&client->subscriptions, subs[i], listEntry);
+            subs[i] = NULL;
```
Why subs[i] = NULL?

wbitesi replied:
In the cleanup part, I add all subs[i] that are not null. the cleanup part is always called independent if there was an error or not. This just ensures that previously added subscriptions are not added again.

I could jump to line 330 to only free the memory and skipping the for loop at the beginning? I'm open to suggestions

wbitesi replied:
Fixed

jpfr commented on src/client/ua_client_subscriptions.c:
```diff
@@ -111,61 +210,134 @@ UA_Client_Subscription_deleteInternal(UA_Client *client, UA_Client_Subscription
         }
 
         UA_Client_Subscription_deleteInternal(client, subs[i]);
+        subs[i] = NULL;
```
subs[i] = NULL?

wbitesi replied:
same as above

jpfr commented on src/client/ua_client_subscriptions.c:
```diff
@@ -111,61 +210,134 @@ UA_Client_Subscription_deleteInternal(UA_Client *client, UA_Client_Subscription
-            UA_LOG_INFO(&client->config.logger, UA_LOGCATEGORY_CLIENT,
-                        "No internal representation of subscription %u",
-                        request.subscriptionIds[i]);
+            UA_LOG_INFO(&client->config.logger, UA_LOGCATEGORY_CLIENT, "No internal representation of subscription %u",
```
Keep lines short.
Avoid cosmetic changes in a feature-commit.

wbitesi replied:
That happened through old clang-format settings. I'm using the new clang-format config and now it doesnt make such changes. Will be fixed later

jpfr commented on src/client/ua_client_subscriptions.c:
```diff
@@ -111,61 +210,134 @@ UA_Client_Subscription_deleteInternal(UA_Client *client, UA_Client_Subscription
+    CustomCallback_remove(cc, false);
+}
+
+UA_StatusCode UA_EXPORT UA_Client_Subscriptions_delete_async(UA_Client *client,
```
UA_EXPORT is not needed in .c file.
Move method name to a new line.

wbitesi replied:
(same thing with clang format)
The existing functions did use UA_EXPORT too, should i remove UA_EXPORT them from my added new functions or from all in ua_client_subscriptions.c?

jpfr replied:
UA_EXPORT is only for headers. Not for .c files.

wbitesi replied:
I have removed all UA_EXPORT occurrences in the ua_client_subscriptions.c file

jpfr commented on src/client/ua_client_subscriptions.c:
```diff
@@ -21,6 +21,47 @@
+cleanup:
+    if (cc->userCallback)
+        cc->userCallback(client, cc->userData, requestId, response);
+    CustomCallback_remove(cc, false);
```
The handler should not have to remove the callback.
This can be done "outside" in the generic handling of async callbacks.

Then, we can also stack-allocate the CustomCallback when call the handler in the synchronous case.

wbitesi replied:
I'm not happy with my approach to delete it directly in the handler either (I previously tried to stack allocate on synchronous functions too). Didnt get it to work properly.

I could add the routine to automatically free a stored customcallback using the requestId in processAsyncMessage (ua_client.c:241). However this would require for every async callback to check (find, delete, remove from list) if there is a customcallback stored, even if there is none saved.

What do you think?

wbitesi replied:
see first comment (CustomCallbackDataDeleter)

jpfr commented on src/client/ua_client_subscriptions.c:
```diff
@@ -283,21 +468,148 @@ __UA_Client_MonitoredItems_create(UA_Client *client,
+    }
+    cc->userCallback = createCallback;
+    cc->userData = userdata;
+    MonitoredItems_CreateData *data = (MonitoredItems_CreateData *)UA_malloc(sizeof(MonitoredItems_CreateData));
```
use UA_calloc.

wbitesi replied:
done

wbitesi commented:
I have updated the pull request according to the comments. I hope the points are addressed well.

Everything tests/builds fine except the clang-analyzer. I don't know why there should be a potential memory leak (which is not shown in the tests). Maybe someone has an idea what i can do to fix this?

Infinity95 commented:
At a glance it looks like you aren't calling the cleanup part if your function returns normally. Maybe you could check if that's the cause.

wbitesi commented:
I have found the thing that clang-analyzer pointed to a memory leak and fixed it with an assertion (with #ifdef __clang_analyzer__). Now everything builds fine.

I have not marked the review as resolved so jpfr can take a look if he's satisfied with the changes, although I think they are now fixed.

wbitesi commented:
I have left the review open for jfp to check if the changes made after his comments are alright.

I have the async functions working in production since April and it works. Is the PR ready for merging or is something still to do?

jpfr commented:
Nice!
This turned out great.
Thanks for the push @wbitesi!
